### PR TITLE
Opt gradient hist mods gain

### LINKED ISSUE #111: GradientBasedOptimizer gradient feature space
--------
Issue Description - FEATURE
--------
##### Which new enhancement should be added?

The normalization of the feature space in the gradient-based optimizers should be done not only for the numerical stability but also for the gradient computation

##### Why?

This enhancement is going to make the opt path much more resilient and will facilitate the optimization parameter settings

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [X] 1. Is it tagged with a type: defect or improvement? ***improvement***
- [X] 2. Is it tagged with a priority: critical, normal or minor? ***critical***
- [X] 3. If it will impact requirements or requirements tests, is it tagged with requirements? ***n/a***
- [X] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. ***n/a***
- [X] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) ***yes***

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed? ***not a defect***
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.) ***not a defect***
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)? ***not a defect***
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)? ***not a defect***
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided? ***#207t***

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #111.

##### What are the significant changes in functionality due to this change request?
 - All data is normalized for internal operations in the optimizer.
 - Step size gain is modified based on the dot product of the last two gradient evaluations, as `a *= g**d`, where `a` is the step size, `g` is the growth factor (set to 2 now, but could be opened to user later), and `d` is the dot product between the previous two gradient evaluations.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/Optimizers/SPSA.py [hunk 1]
```diff
@@ -441,23 +476,72 @@ def _computeGainSequenceCk(self,paramDict,iterNum):
...
+    # modify step size based on the history of the gradients used
+    frac = self.fractionalStepChangeFromGradHistory(traj)
+    ak *= frac
+    self.raiseADebug('step gain size for traj "{}" iternum "{}": {}'.format(traj,iterNum,ak))
+    self.counter['lastStepSize'][traj] = ak
+    return ak
+    # the line search with surrogate unfortunately does not work very well (we use it just at the begin of the search and after that
+    # we switch to a decay constant strategy (above)). Another strategy needs to be find.
+    #### OLD ### skip the line check for now
+    if iterNum > 1 and iterNum <= int(self.limit['mdlEval']/50.0):
```
Comment (alfoa) [inline framework/Optimizers/SPSA.py]: probably we should comment this part of the old code

File: tests/framework/Optimizers/beale.xml [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+  <!-- Test info -->
+  <TestInfo>
+    <name>framework/Optimizers/Beale</name>
+    <author>talbpaul</author>
+    <created>2017-05-03</created>
+    <classesTested>Optimizer</classesTested>
+    <description>
+      This test runs the optimization on Beale's function.  It tests analytic optimization values
+      as well as the mechanical operation of the test.
+    </description>
```
Comment (alfoa) [inline tests/framework/Optimizers/beale.xml]: the tag for ```analytical``` tests is missing

File: framework/Optimizers/Optimizer.py [hunk 3]
```diff
@@ -204,7 +208,29 @@ def _readMoreXMLbase(self,xmlNode):
...
+        #try:
+        #  upper = child.find("upperBound").text
+        #  if lower is None: self.raiseAnError(IOError,"The optional lower and upper bound for the objectVar must be both present at the same time!")
+        #except AttributeError:
+        #  if lower is not None: self.raiseAnError(IOError,"The optional lower and upper bound for the objectVar must be both present at the same time!")
+        #bounds = (utils.floatConversion(lower), utils.floatConversion(upper))
+        #if len(set(bounds)) > 1 and any(v is None for v in bounds):
+        #  self.raiseAnError(IOError,"One of the objecive function bounds is not convertible to a float!")
+        #elif not all(v is None for v in bounds):
+        #  self.objVarBounds = bounds
```
Comment (alfoa) [inline framework/Optimizers/Optimizer.py]: Probabably we should just remove this commented code

**Comment by PaulTalbot-INL** [comment]
I can't make the SPSAwConstraint pass twice consecutively on my machine, let alone multiple test machines.

Any suggestions?  It's batch size 1, and (on my machine) the first 64 entries are the same, then the path diverges and even is different lengths.

**Comment by alfoa** [comment]
seed of the RNG?

**Comment by PaulTalbot-INL** [comment]
It looks like setting the seed isn't good enough.  I'll have to explore options.

**Comment by PaulTalbot-INL** [comment]
It appears tests are passing; awaiting review approval.

**Comment by alfoa** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.  ***Ok***
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). ***Ok***
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. ***Ok***
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. ***They pass***
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. ***Tests added***
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. ***Approve***
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. ***#111***
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added? ***Documentation added***

**Comment by alfoa** [comment]
PR check list passed

**Pull request merged**